jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java

```diff
@@ -3411,9 +3413,54 @@ private static MethodHandle validateArgumentsOfWhileLoop(MethodHandle predHandle
 		Class<?> bodyReturnType = bodyType.returnType;
 		
 		/* The signature of the body handle must be either (V, A...)V or (A...)void */
-		if ((void.class != bodyReturnType) && (bodyReturnType != bodyType.arguments[0])) {
-			/*[MSG "K065D", "The return type of loop body: {0} does not match the type of the first argument: {1}"]*/
-			throw new IllegalArgumentException(com.ibm.oti.util.Msg.getString("K065D", bodyReturnType.getName(), bodyType.arguments[0].getName())); //$NON-NLS-1$
+		int bodyVlength = 0;
+		if (void.class != bodyReturnType) {
+			if (hasNoArgs(bodyType) 
+					|| (bodyReturnType != bodyType.arguments[0])) {
+				String argList = "void";
+				if (! hasNoArgs(bodyType)) {
+					argList = bodyType.arguments[0].getName();
+				}
+				
+				/*[MSG "K065D", "The return type of loop body: {0} does not match the type of the first argument: {1}"]*/
+				throw new IllegalArgumentException(com.ibm.oti.util.Msg.getString("K065D", bodyReturnType.getName(), argList)); //$NON-NLS-1$
+			}
+			bodyVlength = 1;
+		}
+		
+		/* Verify that handles are effectively identical and constrained by body. All we
+		 * need to do here is verify that no parameters have longer argument lists than
+		 * body. The generic loop will compare arguments to ensure all clauses are
+		 * effectively identical.
+		 */
+		int bodyParamTypesLength = bodyHandle.type.arguments.length;
+		int bodyAlength = bodyParamTypesLength - bodyVlength;
+		
+		/* predHandle parameters are empty or (V, A..). Make sure it is compatible with
+		 * the effectively identical model that compares A. An empty list can be skipped
+		 * as it is always effectively identical.
+		 */
+		MethodType predParamType = predHandle.type;
+		if (! hasNoArgs(predParamType)) {
+			Class<?>[] predParamTypes = predHandle.type.arguments;
+			
+			if (void.class != bodyReturnType) {
+				if (bodyReturnType != predParamTypes[0]) {		
+					/*[MSG "K0654", "Pred parameter list must be either empty or of the form (V A*)"]*/		
+					throw new IllegalArgumentException(com.ibm.oti.util.Msg.getString("K0654")); //$NON-NLS-1$		
+				}
+			}
+			
+			int externalPredParamTypesLength =  predParamType.parameterCount() - bodyVlength;
+			if (externalPredParamTypesLength > bodyAlength) {
```

Comment:
Unless Ive missed it, it looks like this only checks that you don't have more `A` params in the predicate than in the loop body, but I think you also need to check that the types match up. 
  
ie. 
```
		MethodHandle init = lookup.findStatic(MethodHandlesTest.class, "init", MethodType.methodType(int.class, long.class));
		MethodHandle pred = lookup.findStatic(MethodHandlesTest.class, "pred", MethodType.methodType(boolean.class, int.class, int.class));
		MethodHandle body = lookup.findStatic(MethodHandlesTest.class, "body", MethodType.methodType(int.class, int.class, long.class));
```

That should fail with `loop predicate must match: (int,int)boolean != (int,long)boolean`.

Reply:
similar comment for the init handle

Reply:
I only checked the lengths instead of looping through all the parameters in `doWhileLoop`, `whileLoop`, and `countedLoop` because they pass all their parameters to the generic loop. My reasoning was that the generic loop (`loop()`) checks that each MethodHandle is effectively identical to the longest MethodHandle thats passed in. As long as body is at least as long as init or pred the generic loop should be sufficient.

I'll work on matching up the error message.

Reply:
Ah but then the error message won't be correct.. I'll add this in.

Reply:
This is fixed in the latest commit.